### ISSUE TITLE
Envoy's L4 Golang extension

### ISSUE THREAD

**Issue by antJack** [envoyproxy/envoy#25906]
Recently, my colleague implemented the [L7 golang extension](https://github.com/envoyproxy/envoy/pull/22573). Similarly, we have implemented an L4 golang extension using the same cgo technique as L7. And the L4 extension has already been widely utilized within our company for almost a year, enabling our internal customized protocol to benefit from the high-performance network infrastructure of Envoy and the superior development efficiency of Golang. Additionally, the L4 golang extension can also provide a better implementation path for envoy's future support of Dapr, etc. Given the above, we are consider to open source our L4 golang extension very soon.

**Comment by taoyuanyuan** [comment]
Cool stuff !~

**Comment by rakesh-eltropy** [comment]
I tired to use this extension.

```
func (f *filter) DecodeHeaders(header api.RequestHeaderMap, endStream bool) api.StatusType {

	f.path, _ = header.Get(":path")
	fmt.Println(f.path)

	method, _ := header.Get(":method")
	fmt.Println(method)

	useragent, _ := header.Get(":useragent")
	fmt.Println(useragent)

	host, _ := header.Get(":host")
	fmt.Println(host)
	

	if f.path == "/localreply_by_config" {
		return f.sendLocalReplyInternal()
	}
	return api.Continue
}
```
I am able to get path and method but other headers are blank. Also, I see header.Range is still not implemented so not sure what all keys are available. Can you please guide me hos to get host, useragent etc headers.

**Comment by doujiang24** [comment]
@rakesh-eltropy This issue is for L4 filter, while your question is L7 filter related, better create a new issue for it.

For your questions:
1. for the keys of the header, you could follow this, i.e. `header.Get("user-agent")`.
https://github.com/envoyproxy/envoy/blob/main/source/common/http/headers.h#L212
2. we plan to introduce some new methods in the HeaderMap interface, i.e. `Host`, `Path`, `Method` and etc. so that people could use it easier.

Yep, `Range` and doc are on the way, thanks for your try and reminder.

**Comment by antJack** [comment]
https://github.com/envoyproxy/envoy/pull/26173

**Comment by huanghuangzym** [comment]
> #26173

when will support l4 extenstion?

**Comment by gitanuj** [comment]
I see this has landed in main recently. Is there any plan to support reading/writing filter state and dynamic metadata from it? This looks perfect for a custom routing filter at L4 that I'm planning to work on.

**Comment by antJack** [comment]
> I see this has landed in main recently. Is there any plan to support reading/writing filter state and dynamic metadata from it? This looks perfect for a custom routing filter at L4 that I'm planning to work on.

@gitanuj can you please provide more details about your requirements? so that we could evaluate its feasibility. Of course, you are also welcome to submit a PR yourself.

**Comment by gitanuj** [comment]
@antJack I actually created an issue about it #28128 

To elaborate - I'm using Istio Ingress Gateway and I want to route TCP connections to different kubernetes services based on PPv2 TLV Headers (specifically the VPC Endpoint ID which is used for Private Links). I haven't found any existing solution for it so here's what I'm planning:
1. Use [ProxyProtocol listener filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/proxy_protocol) (which parses and emits TLV headers as dynamic metadata, there's also an option to passthrough TLV headers as filter state)
2. Write a custom network filter (using this L4 Golang extension) and use the TLV header value from either dynamic metadata or filter state and set `envoy.tcp_proxy.cluster` on filter state object
3. Use [TCP Proxy network filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/tcp_proxy_filter) to route to an upstream cluster (Istio creates outbound clusters for every kubernetes service automatically). This filter uses `envoy.tcp_proxy.cluster` filter state to route to a specific cluster

Does this make sense? I would be happy to work on the change too.

**Comment by antJack** [comment]
@gitanuj sound good. 

The golang l7 extension currently has the capability to get or set dynamic metadata and filter state, and the l4 extension will soon have these features as well.

**Comment by honeybadger10** [comment]
I'm looking into building a custom filter to authorize at L4 and use the TCP Proxy network filter to route to an upstream cluster
@antJack are there still plans to add `continueReading()` to allow chaining the L4 Go Extension with the tcp_proxy filter and also preventing the filter from draining the connection buffer, so the information is still available for the next filter?

**Comment by doujiang24** [comment]
@honeybadger10 yep, it's on the TODO list, but not scheduled for the short term.

**Comment by rileyp2000** [comment]
I have a need to use the golang network filter as a non terminal filter (i.e. before the tcp_proxy filter or another that handles routing to upstream clusters). It seems like there is still some open work in this area to get it up and running. I'd be happy to contribute here, but was curious if anyone had more info about the necessary changes for the effort. When compared with the HTTP filter it seems like there are some more significant differences between the go interfaces and the c++ code for a typical network filter. Thanks!

**Comment by agrawroh** [comment]
@rileyp2000 We just added support for [Dynamic Modules](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/dynamic_modules) for [Network Filters](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/dynamic_modules/v3/dynamic_modules.proto#envoy-v3-api-msg-extensions-filters-network-dynamic-modules-v3-dynamicmodulenetworkfilter).

Just like HTTP, we can support Terminal Network Filters. You think that's something which would work for you? We are happy to take feedback on anything missing from the ABI and add support for it.

cc @wbpcode @mathetake

**Comment by rileyp2000** [comment]
@agrawroh Thanks for that info! I must have missed the announcement about dynamic modules getting introduced a few versions ago. Now that I am aware :), I have a couple questions if you or someone else is able to answer them. Are there plans to sunset the golang filter(s) and replace it with dynamic modules? Also, we currently use the HTTP golang filter, is the ABI stable enough that we should look into migrating our work from there to dynamic modules?

I've been looking at the examples repo which has been super helpful in understanding the ABI. To clarify, what we are looking for is a non-terminal network filter, which would then pass the connection on to something like the TCP proxy filter for cluster routing. Is that currently supported with the ABI and the go sdk in the examples repository? I wasn't able to find some of the network filter functions implemented in the go sdk. I'd be happy to contribute there!

**Comment by agrawroh** [comment]
@rileyp2000 Usually the development and feature requests goes to the extension owners. Unfortunately, I don't really know the current owners of GoLang extension(s) and can't commit to their availability in adding newly requested features but there is no plans to sunset it. 

That said, Dynamic Modules are currently actively being used by us at Databricks and also other workplaces. I certainly will be interested in adding new features as that would also make it easier for us to adopt it for more use-cases.

I'd recommend you to take a look at the ABI (it just got added in the recent version for Network Filter) and see if it would meet your current use-cases. If not, we are happy to extend the ABI as needed.

**Comment by doujiang24** [comment]
> I have a need to use the golang network filter as a non terminal filter (i.e. before the tcp_proxy filter or another that handles routing to upstream clusters). It seems like there is still some open work in this area to get it up and running. I'd be happy to contribute here, but was curious if anyone had more info about the necessary changes for the effort. When compared with the HTTP filter it seems like there are some more significant differences between the go interfaces and the c++ code for a typical network filter. Thanks!

Hi @rileyp2000 , sorry for the late, I'm happy to review the PR, but I don't have much time to implement it.
